ENH: added plot_event function

robbisg opened the pull request:
This PR should fix #2385

robbisg commented:
I have to add import in ```__init__.py```

bthirion commented:
You need to provide tests ;-)

robbisg commented:
> You need to provide tests ;-)

Sure?? :rofl:

bthirion commented on nilearn/reporting/_plot_matrices.py:
```diff
@@ -62,6 +62,87 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None, output_file=None):
+
+    Parameters
+    ----------
+    model_event : DataFrame or list of DataFrame
```
If possible, could you adopt the best conventions for docstrings, as in #2450

bthirion commented:
Also, it would be good to showcase it in examples such as 
https://github.com/nilearn/nilearn/blob/master/examples/04_glm_first_level_models/plot_spm_multimodal_faces.py
or
https://github.com/nilearn/nilearn/blob/master/examples/plot_single_subject_single_run.py

robbisg commented:
Great! Thank you!

robbisg commented:
I will add showcase and fix docsting, next.

GaelVaroquaux commented on nilearn/reporting/_plot_matrices.py:
```diff
@@ -62,6 +62,87 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None, output_file=None):
+
+    """
+    import matplotlib.patches as mpatches
+    import pandas as pd
```
Is there a reason why those imports are here, rather than at the module level?

robbisg replied:
I forgot to remove it! I will add it at module level!

GaelVaroquaux commented on nilearn/reporting/_plot_matrices.py:
````diff
@@ -62,6 +62,87 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None, output_file=None):
+    Parameters
+    ----------
+    model_event : DataFrame or list of DataFrame
+        the list of Dataframe obtained by ```first_level_models_from_bids```
````
```suggestion
        the list of Dataframe obtained for instance by :func:`~nilearn.stats.first_level_model.first_level_models_from_bids`
```
Also, I think that it would be useful to specify what the importance columns of the data-frame are.

robbisg replied:
OK!

GaelVaroquaux commented on nilearn/reporting/_plot_matrices.py:
```diff
@@ -62,6 +62,87 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None, output_file=None):
+    cmap : str or matplotlib.cmap, optional
+        the colormap used to label different events
+
+    output_file: string or None, optional,
```
```suggestion
    output_file : string or None, optional,
```

robbisg commented:
> My additional important comment would be that the new function should be added to:
> 
> * An existing example to demo it
I was thinking about extending the ```examples/04_glm_first_level_models/write_events_file.py``` and make it more "complicated", by using a list of event files, for example. What do you think ?

bthirion commented:
I like the suggestion, but wouldn't try to make the example more complicated ;-)

GaelVaroquaux commented:
>         I was thinking about extending the examples/04_glm_first_level_models/write_events_file.py and make it more "complicated", by using a list of event files

Can you not just add the function there without extending it?

robbisg commented:
@GaelVaroquaux Sure!
Basically, my idea was to reduce the complexity of the example since, as discussed in the issue page with @bthirion, this is already an extreme case with more than 10 events!!

GaelVaroquaux commented:
> Basically, my idea was to reduce the complexity of the example since, as discussed in the issue page with @bthirion, this is already an extreme case with more than 10 events!!

Ah, sorry, I had missed that. Reducing the complexity seems a good idea!

robbisg commented:
> I made a few comments, mostly minor.

I've refactored the function since it appeared bad with more than 5 runs (e.g. in Haxby example)

> 
> My additional important comment would be that the new function should be added to:
> 
> * An existing example to demo it

I added it to ```write_events_file.py``` which I renamed it in ```plot_events_file.py``` and also in ```plot_haxby_block_classification.py``` since it may be used also to address #2352 .

> * The file doc/modules/reference.rst so that it appears in the reference documentation

Done!

> * The documentation doc/plotting.rst

I don't know in which section add ```plot_event```, I can also make a section for ```nilearn.reporting``` plot function, if you'd like.

I included the fix of #2468 in this PR

bthirion commented on examples/04_glm_first_level_models/plot_events_file.py:
```diff
@@ -1,5 +1,5 @@
+=======================================================================================
 
 The protocol described is the so-called "archi standard" localizer
 event sequence.  See Pinel et al., BMC neuroscience 2007 for reference.
```
Maybe add that  this is a slightly simplified version of the protocol.

bthirion commented on nilearn/reporting/_plot_matrices.py:
```diff
@@ -62,6 +64,94 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None, output_file=None):
+    else:
+        cmap = cmap
+
+    event_labels = pd.concat(e['trial_type'] for e in model_event)
```
```suggestion
    event_labels = pd.concat(event['trial_type'] for event in model_event)
```Bzw, I'm surprise that this works. Thought that we would need `[event['trial_type'] for event in model_event]`

robbisg replied:
I read it few days ago from a colleague and it worked :)!!
I think that In some cases it is more readable but it is not straightforward that you are passing a list!

GaelVaroquaux commented on doc/install_doc_component.html:
```diff
@@ -118,8 +118,9 @@
-      (sometimes called <cite>python-nibabel</cite>).</p>
+      called <cite>python-matplotlib</cite>), <strong>nibabel</strong>
+      (sometimes called <cite>python-nibabel</cite>) and <strong>pandas</strong>
+      (sometimes called <cite>python-pandas</cite>).</p>
```
These look like changes from another PR. I probably have a mangled history.

Are you working from a branch, and not master? To contribute to software, you should always make sure to work in branches, in order to avoid mixing your changes.

robbisg replied:
I was working on nistats_plot_event branch, but I added these changes in this PR since in #2468 others told me to do not open another PR, in order to save CI runs. 
So I decided to add these fix to this branch and this PR.
I don't know if something occurred when I rebased master to this branch or something else.

GaelVaroquaux replied:
Sounds good.

I'll merge once the suggestions are addressed.

GaelVaroquaux commented on nilearn/reporting/_plot_matrices.py:
```diff
@@ -62,6 +64,94 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None, output_file=None):
+        patch = mpatches.Patch(color=cmap.colors[idx], label=label)
+        handles.append(patch)
+
+    _ = ax.legend(handles=handles)
```
As visible on https://5706-1235740-gh.circle-artifacts.com/0/doc/_build/html/auto_examples/04_glm_first_level_models/plot_events_file.html#sphx-glr-auto-examples-04-glm-first-level-models-plot-events-file-py
I think that the plot would be nicer if you increased a bit the number of cols in the legend, using "ncols=3" or 4.

robbisg replied:
Ok!

GaelVaroquaux commented:
I fixed the conflict, pull this branch before you work on it again.

GaelVaroquaux commented:
LGTM; Will merge when CI is green.

tsalo commented:
Pinging @GaelVaroquaux. CI passed. Is this good to merge?

GaelVaroquaux commented:
Yes this is good for merge. I'm merging this. Thanks a lot for the ping @tsalo

bthirion commented:
Thx for adding this feature !
